Implement ability to load images by default in non-Docker build drivers.

n-g opened the pull request:
This eases build driver migrations, as it allows aligning the default behavior.

See also https://docs.docker.com/build/drivers/

tonistiigi commented on controller/build/build.go:
```diff
@@ -133,6 +133,15 @@ func RunBuild(ctx context.Context, dockerCli command.Cli, in controllerapi.Build
 
+	// ENV-guarded ability load images by default if no other output is specified.
+	// This is useful when migrating from Docker buildx drivers to others, as it aligns the default behavior.
+	if build.DefaultLoad() && len(outputs) == 0 && len(opts.Tags) > 0 {
```
why the `opts.Tags` check in here?

n-g replied:
My thinking is:
Loading an image from a remote driver can be slow as networking is involved.
So, I'd like to differentiate the case where I just want to check if a build passes (and populate the cache of the remote build driver) vs I tag the image so that I can reuse it later.

tonistiigi reviewed (commented):
This could also be a `driver-opt` maybe?

n-g commented:
> This could also be a `driver-opt` maybe?

Thanks! I like this path, as it allows finer granularity when selecting the new behavior (per driver vs per env).
From what I can tell, this is the first `driver-opt` applicable to multiple driver types.
I introduced a new feature, but lmk if there is a cleaner modeling path I could employ.

crazy-max commented on driver/docker-container/driver.go:
```diff
@@ -44,18 +44,19 @@ type Driver struct {
+	cgroupParent      string
+	restartPolicy     container.RestartPolicy
+	env               []string
+	defaultLoadTagged bool
```
I think `defaultLoad` is enough to be consistent with `BUILDX_NO_DEFAULT_LOAD` env var: https://docs.docker.com/build/building/variables/#build-tool-configuration-variables

n-g replied:
Thanks! Only loading tagged images was supposed to be a latency optimization as a pull from a remote driver can be slow.
But I agree that consistency is important. Updated.

crazy-max commented on driver/features.go:
```diff
@@ -7,3 +7,5 @@ const DockerExporter Feature = "Docker exporter"
 const CacheExport Feature = "Cache export"
 const MultiPlatform Feature = "Multi-platform build"
+
+const DefaultLoad Feature = "Automatically load images"
```
```suggestion
const DefaultLoad Feature = "Automatically load images to docker store"
```

@dvdksn for copy :pray:

n-g commented:
@tonistiigi @crazy-max
Thanks for taking a look at this!
From what I can tell, I addressed all your suggestions.
Is there anything left to do to make this PR ready to merge?
Thanks!

dvdksn commented on driver/features.go:
```diff
@@ -7,3 +7,5 @@ const DockerExporter Feature = "Docker exporter"
 const CacheExport Feature = "Cache export"
 const MultiPlatform Feature = "Multi-platform build"
+
+const DefaultLoad Feature = "Automatically load images to docker store"
```
```suggestion
const DefaultLoad Feature = "Automatically load images to the Docker Engine image store"
```

n-g replied:
done - thanks!

crazy-max commented on driver/features.go:
```diff
@@ -7,3 +7,5 @@ const DockerExporter Feature = "Docker exporter"
 const CacheExport Feature = "Cache export"
 const MultiPlatform Feature = "Multi-platform build"
+
+const DefaultLoad Feature = "Automatically load images to the Docker Engine image store"
```
@dvdksn I think "Docker store" is enough no?

```suggestion
const DefaultLoad Feature = "Automatically load images to the Docker store"
```

dvdksn replied:
"Docker store" was the name of an old publisher tool of Docker Hub so that term becomes a little unfortunate for that reason. In general we always use "image store", and I think that keeping "Docker Engine" in there makes it clearer that we're referring to the local content store for the daemon.

crazy-max reviewed (commented):
Can you squash your commits please?

Also would be nice to have integration tests in https://github.com/docker/buildx/tree/master/tests if you can or can be follow-up

Thanks!

n-g commented:
> Can you squash your commits please?
> 
> Also would be nice to have integration tests in https://github.com/docker/buildx/tree/master/tests if you can or can be follow-up
> 
> Thanks!

Sure! 

I initially did not realize that not all workers are tested by default with `./hack/test`.
Updated the test now, and validated it with
`TESTFLAGS="-v --parallel=1 --timeout=30m --run=//worker=docker$" TEST_DOCKERD=1 TESTFLAGS_DOCKER="-v --parallel=1 --timeout=30m" ./hack/test` locally.

PTAL

crazy-max commented on tests/build.go:
```diff
@@ -675,6 +676,69 @@ func testBuildLoadPush(t *testing.T, sb integration.Sandbox) {
+		"--buildkitd-flags=--allow-insecure-entitlement=network.host",
+		"--driver-opt", "network=host",
+		"--driver-opt", "default-load=true",
+		"--driver-opt", "image=moby/buildkit:v0.13.0-rc3",
```
not needed for this test iiuc

n-g replied:
Thx, simplified to only set the driver-opt that we're testing.

crazy-max commented on tests/build.go:
```diff
@@ -675,6 +676,60 @@ func testBuildLoadPush(t *testing.T, sb integration.Sandbox) {
+		require.NoError(t, err, out)
+	})
+
+	// TODO: use stable buildkit image when v0.13.0 released
```
remove this as well

crazy-max commented on tests/build.go:
```diff
@@ -675,6 +676,60 @@ func testBuildLoadPush(t *testing.T, sb integration.Sandbox) {
+	cmd.Stderr = os.Stderr
+	require.NoError(t, cmd.Run())
+
+	// TODO: test metadata file when supported by multi exporters https://github.com/docker/buildx/issues/2181
```
remove this

n-g commented:
> @crazy-max Could you help me re-trigger the validation workflows?

Thanks!
Let me know if there is anything else I can help with here 🙏

crazy-max reviewed (dismissed):
LGTM thanks!

crazy-max commented on tests/build.go:
```diff
@@ -675,6 +676,57 @@ func testBuildLoadPush(t *testing.T, sb integration.Sandbox) {
 }
 
+func testBuildDefaultLoad(t *testing.T, sb integration.Sandbox) {
+	if sb.Name() != "docker" {
+		t.Skip("skipping test for non-docker workers")
+	}
```
Use helper, see https://github.com/docker/buildx/commit/bcda1dbcdfafb36c558aa4a550bf14cbd7a93ef5

crazy-max commented on tests/build.go:
```diff
@@ -675,6 +676,57 @@ func testBuildLoadPush(t *testing.T, sb integration.Sandbox) {
+		t.Skip("skipping test for non-docker workers")
+	}
+
+	registry, err := sb.NewRegistry()
+	if errors.Is(err, integration.ErrRequirements) {
+		t.Skip(err.Error())
+	}
+	require.NoError(t, err)
+
+	target := registry + "/buildx/registry:" + identity.NewID()
```
Setting up registry is not needed for this tests, see https://github.com/docker/buildx/commit/4d012cfa1f3f9d9037483773919c355961fe38f1

crazy-max reviewed (changes requested):
It needs a rebase with changes introduced recently sorry :sweat_smile:
See also review, I tried to push changes to your branch but looks like you don't allow edits by maintainers. Here is my branch if you want to pick these changes: https://github.com/docker/buildx/compare/master...crazy-max:buildx:default-load-opt otherwise I can carry in another PR if you prefer. Thanks

n-g commented:
> It needs a rebase with changes introduced recently sorry 😅 See also review, I tried to push changes to your branch but looks like you don't allow edits by maintainers. Here is my branch if you want to pick these changes: [master...crazy-max:buildx:default-load-opt](https://github.com/docker/buildx/compare/master...crazy-max:buildx:default-load-opt) otherwise I can carry in another PR if you prefer. Thanks

Thanks!
Rebased and applied your suggestions.

From what I can tell, I cannot grant edit rights retroactively.
Feel free to ping me if you need any changes.

crazy-max commented on tests/build.go:
```diff
@@ -750,6 +751,46 @@ COPY --from=build /token /
+		fmt.Sprintf("-t=%s", tag),
+		dir,
+	))
+	cmd.Env = append(cmd.Env, "BUILDX_BUILDER="+builderName)
+	outb, err := cmd.CombinedOutput()
+	require.NoError(t, err, string(outb))
+}
```
The `image inspect` test seems to have been removed: https://github.com/docker/buildx/compare/master...crazy-max:buildx:default-load-opt#diff-12855e5a65cd0dc084324f4808d159cf2ebd514a77c6eee4dba653e1a4e94470R793-R795

n-g replied:
Yup, sorry, bad merge. Was juggling too many things.
Updated and tested locally 🤞

crazy-max reviewed (approved):
Thanks!
